Removing SSLv2 and SSLv3 support from System.Net.Security

CIPop opened the pull request:
APIs will throw NotSupportedException if SSLv2 or v3 is used. This behavior is different from .NET Desktop.
Tests have been changed to pin this new behavior.
Increased passing-test timeouts to 15s.

Fix #3114 (partial), #4467

@davidsh @josguil @bartonjs @vijaykota @rajansingh10 @shrutigarg PTAL
/cc: @stephentoub @SidharthNabar

davidsh commented on src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs:
```diff
@@ -0,0 +1,219 @@
+
+        public ServerAsyncAuthenticateTest()
+        {
+            _log = TestLogging.GetInstance();
```
What is the difference between using `TestLogging.GetInstance()` and using the constructor overload which passes in an `ITestOutputHelper` instance?  We have other tests that use the latter.

CIPop replied:
The default XUnit ITestOutputHelper doesn't work if called outside the context of a test. 

TestLogging is System.Net specific and uses ETW for fast and reliable logging regardless of context. See https://github.com/dotnet/corefx/blob/master/Documentation/debugging/windows-instructions.md for information on how to collect these traces.

davidsh replied:
Ok.  I understand now. I was confused because I saw the Xunit interface type `ITestOutputHelper`. But it appears that our System.Net ETW tracing supports that interface.

stephentoub commented:
> APIs will throw NotSupportedException if SSLv2 or v3 is used. This behavior is different from .NET Desktop.

For my own edification, this is being done because these protocols are outdated, insecure, etc.?

stephentoub commented on src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs:
```diff
@@ -460,5 +468,15 @@ public override void Write(byte[] buffer, int offset, int count)
+
+        private static void ValidateSecurityProtocol(SslProtocols protocols)
+        {
+            SslProtocols allowedProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
```
This can be ```const```

CIPop replied:
Will extract in a const member and place in Common to unify validation code. I'll try to change other places where protocols are filtered.

davidsh commented:
>For my own edification, this is being done because these protocols are outdated, insecure, etc.?

SSL2 and SSL3 are not safe protocols anymore. The security guidance for new software releases is to make them more secure in general. The same guidance was used in WinHttpHandler to not support SSL2 and SSL3.

stephentoub commented:
LGTM

akoeplinger commented:
It'd be great if that reasoning could be added to the exception message so users don't go hunting for ways to enable them.

SidharthNabar commented:
@roncain @vijaykota @saurabh500 @danroth27 - PTAL if this impacts any of your scenarios.

Some comments - 
1. +1 to @akoeplinger 's comment about also including why they are not supported in the exception message.

2. Is there a way for developers to enable these protocols for .NET Core in any way? An opt-in switch, etc? If yes, we should document that method. If not, state that in the description of this issue so that the impact is clear.

3. What are the APIs whose callers are impacted by this? We should list that in the top-level description of the issue as well. I see SslStream.AuthenticateAsClient and SslStream.AuthenticateAsServer being affected from the commit. Any others?

CIPop commented:
> It'd be great if that reasoning could be added to the exception message so users don't go hunting for ways to reenable them.

@SidharthNabar @akoeplinger The message was copied from .Net Desktop. This exception is thrown when someone tries to set Ssl2 as `ServicePointManager.SecurityProtocol`:

```C# 
               throw new NotSupportedException(SR.GetString(SR.net_securityprotocolnotsupported));
```

I don't think it's a good idea to change this and I don't think we ever gave reasons on why we don't support other protocols like PCT 1.0 (or crypto algorithms such as RC4, MD5) outside of security KB articles. 

If we would change it, what are the proposals and how would they scale when new protocols are excluded? (e.g. We already silently refuse TLS1.0 + RC4 and many other insecure crypto/hashing cyphersuites as documented in various security KBs released with .Net 4.6. This list is likely to grow and change fast, as soon as a protocol or cyphersuite is found insecure.)

CIPop commented:
@SidharthNabar 

> Is there a way for developers to enable these protocols for .NET Core in any way? An opt-in switch, etc? If yes, we should document that method. If not, state that in the description of this issue so that the impact is clear.

Updated the PR comment. I've already left a comment within the code.

> What are the APIs whose callers are impacted by this? We should list that in the top-level description of the issue as well. I see SslStream.AuthenticateAsClient and SslStream.AuthenticateAsServer being affected from the commit. Any others?

Within CoreFX another affected component is `System.Data.SqlClient`.

davidsh commented:
LGTM

CIPop commented:
@davidsh @stephentoub @vijaykota PTAL

I've made changes to unify the Security Protocol selection within SslStream and WinHttpHandler. 
Added unit tests for SslStream and modified existing tests and behavior for WinHttpHandler (now throws NotSupported exception instead of ArgumentException).

davidsh commented:
Added some comments/questions on the new commit.

davidsh commented:
CI failed with compilation errors:
`error CS0117: 'System.SR' does not contain a definition for 'net_securityprotocolnotsupported' [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Http\src\System.Net.Http.csproj`

You need to add the new string resource to both WinHttpHandler and Http libraries (due to the weirdness of how they get built)

shrutigarg commented on src/System.Net.Security/tests/FunctionalTests/CertificateValidationClientServer.cs:
```diff
@@ -69,19 +68,19 @@ public CertificateValidationClientServer()
                     Task clientAuthentication = sslClientStream.AuthenticateAsClientAsync(
                         serverName,
                         clientCerts,
```
is this test passing in Linux? client cert validation passes without any prerequisite step?

CIPop replied:
Assuming that CI runs them, all tests have passed. What was the prerequisite step that you would have expected here?

shrutigarg replied:
Stephen had disabled the Linux runs again in CI.. So they wouldn't have run on Linux. 
For client certificate validation ,CRL check is enabled by default in Unix implementation and I think the Cert which we are passing will not be validated until we have CA available all the time? 
How does it work in WIndows?
Actually the test is failing in the cert validation part only in unix (I ran them locally)

shrutigarg commented on src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs:
```diff
@@ -0,0 +1,197 @@
+
+                    Task serverAuthentication = sslServerStream.AuthenticateAsServerAsync(
+                        _serverCertificate,
+                        true,
```
Shouldn't this be false?
 we are not passing any client certificate above in AuthenticateAsClientAsync.?
Were these tested in Unix?

CIPop replied:
This is a good point. I'll investigate why this allows some of the connections.

CI passing is a requirement before merge and all tests were marked as passing.

shrutigarg replied:
not sure how windows behave in this case .. may be it ignores and continues if client doesn't pass a cert. 
In Unix it fails. And as said above Unix tests are not run currently in CI.
